src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirOperationDataStore.cs

```diff
@@ -195,9 +194,147 @@ public async Task<IReadOnlyCollection<ExportJobOutcome>> AcquireExportJobsAsync(
+                }
+
+                return new ReindexJobWrapper(jobRecord, WeakETag.FromVersionId(rowVersion.ToString()));
+            }
+        }
+
+        public async Task<ReindexJobWrapper> GetReindexJobByIdAsync(string id, CancellationToken cancellationToken)
+        {
+            EnsureArg.IsNotNullOrWhiteSpace(id, nameof(id));
+
+            using (SqlConnectionWrapper sqlConnectionWrapper = await _sqlConnectionWrapperFactory.ObtainSqlConnectionWrapperAsync(cancellationToken, true))
+            using (SqlCommandWrapper sqlCommandWrapper = sqlConnectionWrapper.CreateSqlCommand())
+            {
+                VLatest.GetReindexJobById.PopulateCommand(sqlCommandWrapper, id);
+
+                using (SqlDataReader sqlDataReader = await sqlCommandWrapper.ExecuteReaderAsync(CommandBehavior.SequentialAccess, cancellationToken))
+                {
+                    if (!sqlDataReader.Read())
+                    {
+                        throw new JobNotFoundException(string.Format(Core.Resources.JobNotFound, id));
+                    }
+
+                    (string rawJobRecord, byte[] rowVersion) = sqlDataReader.ReadRow(VLatest.ReindexJob.RawJobRecord, VLatest.ReindexJob.JobVersion);
+
+                    return CreateReindexJobWrapper(rawJobRecord, rowVersion);
+                }
+            }
+        }
+
+        public async Task<ReindexJobWrapper> UpdateReindexJobAsync(ReindexJobRecord jobRecord, WeakETag eTag, CancellationToken cancellationToken)
+        {
+            EnsureArg.IsNotNull(jobRecord, nameof(jobRecord));
+
+            byte[] rowVersionAsBytes = GetRowVersionAsBytes(eTag);
+
+            using (SqlConnectionWrapper sqlConnectionWrapper = await _sqlConnectionWrapperFactory.ObtainSqlConnectionWrapperAsync(cancellationToken, true))
+            using (SqlCommandWrapper sqlCommandWrapper = sqlConnectionWrapper.CreateSqlCommand())
+            {
+                VLatest.UpdateReindexJob.PopulateCommand(
+                    sqlCommandWrapper,
+                    jobRecord.Id,
+                    jobRecord.Status.ToString(),
+                    JsonConvert.SerializeObject(jobRecord, _jsonSerializerSettings),
+                    rowVersionAsBytes);
+
+                try
+                {
+                    var rowVersion = (byte[])await sqlCommandWrapper.ExecuteScalarAsync(cancellationToken);
+
+                    if (rowVersion.NullIfEmpty() == null)
+                    {
+                        throw new OperationFailedException(string.Format(Core.Resources.OperationFailed, OperationsConstants.Reindex, "Failed to create reindex job because no row version was returned."), HttpStatusCode.InternalServerError);
+                    }
+
+                    return new ReindexJobWrapper(jobRecord, GetRowVersionAsEtag(rowVersion));
+                }
+                catch (SqlException e)
+                {
+                    if (e.Number == SqlErrorCodes.PreconditionFailed)
```

Comment:
TODO: instead of throwing, investigate adding a delay, getting the updated job and retrying

Reply:
Investigated and created [AB#80852](https://microsofthealth.visualstudio.com/Health/_workitems/edit/80852), which will follow this PR.